Add `layout.shapes.layer`

n-riesco opened the pull request:
* Added functionality to show shapers below traces.

* The axis-references of a 'below'-shape determine the subplots under which a shape is shown.

* If both axis-references of a 'below'-shape are set to 'paper', then the shape is shown below all the subplots.

* Updated `test/image/mocks/shapes.json` to exercise the new functionality.

* Updated `test/jasmine/tests/shapes_test.js` to account for the new shape layers.

Closes #148

n-riesco commented:
This PR is ready for review.

mdtusz commented:
Seems reasonable to me - we generally like to put have a bit of discussion when adding attributes - thoughts on this @etpinard?

I still have a few things I'd like to check on this, but it looks good so far. Maybe take a look into resolving whatever the conflict is and create an image test for the two functionalities as well - would be useful for catching future shapelayer bugs.

etpinard commented:
@n-riesco Looks like you're on the right track.

> The axis-references of a 'below'-shape determine the subplots under which a shape is shown.

Agreed. :+1: . But I found one bug. See below comment.

> If both axis-references of a 'below'-shape are set to 'paper', then the shape is shown below all the subplots

You mean if `xref` and `yref` are set to `'paper'` ? In that case, yep, I agree :+1:

etpinard commented:
Consider: 


```js
var y = [1,2,1,0,-1,2,3,5];

Plotly.plot(Tabs.fresh(), [
    {
        y: y,
        line: { shape: 'spline' },
	}, {
        y: y.map((v) => 10 * Math.sin(Math.PI * v / 4)),
        line: { shape: 'spline' },
        xaxis: 'x2'
    }, {
        y: y.map((v) => 10 * Math.cos(Math.PI * v / 4)),
        line: { shape: 'spline' },
        yaxis: 'y2'
    }, {
        y: y.map((v) => 10 / (v + 4)),
        line: { shape: 'spline' },
        xaxis: 'x2',
        yaxis: 'y2'
    }],
    {
        title: 'shape shading a region',
        showlegend: false,
        xaxis: {
            domain: [0, 0.45]
        },
        xaxis2: {
            domain: [0.55, 1]
        },
        yaxis: {
            domain: [0, 0.45]
        },
        yaxis2: {
            domain: [0.55, 1]
        },
        shapes: [ 
            {
                type: 'rect',
                layer: 'below',
                xref: 'x',
                x0: 3.5,
                x1: 4.5,
                yref: 'paper',
                y0: 0,
                y1: 1,
                fillcolor: '#c7eae5',
            },
            {
                type: 'rect',
                layer: 'above',
                xref: 'x2',
                x0: 5.5,
                x1: 6.5,
                yref: 'paper',
                y0: 0,
                y1: 1,
                fillcolor: '#c7eae5',
                opacity: 0.5
            },
            {
                type: 'rect',
                layer: 'below',
                xref: 'paper',
                x0: 0,
                x1: 1,
                yref: 'y',
                y0: 0,
                y1: 3,
                fillcolor: '#f6e8c3',
            },
            {
                type: 'rect',
                layer: 'above',
                xref: 'paper',
                x0: 0,
                x1: 1,
                yref: 'y2',
                y0: 1,
                y1: 4,
                fillcolor: '#f6e8c3',
                opacity: 0.5
            },

            {
                type: 'rect',
                layer: 'below',
                xref: 'paper',
                x0: 0.3,
                x1: 0.7,
                yref: 'paper',
                y0: 0.3,
                y1: 0.7,
                fillcolor: '#d3d3d3',
            }
        ],
        dragmode: 'pan'
    }
)
```

This gives 

![image](https://cloud.githubusercontent.com/assets/6675409/14618078/ddd42fe4-057e-11e6-815f-c2b04ef090ab.png)

where both the `xref: 'x'` and the `yref: 'y'`  shapes appear above the `xy` subplot.

etpinard commented:
TODOs:

- [ ] fix issue presented https://github.com/plotly/plotly.js/pull/439#issuecomment-211552620
- [ ] add one image test (example in https://github.com/plotly/plotly.js/pull/439#issuecomment-211552620 would be ok, or something even more comprenhensive)
- [ ] add one relayout test where a shape `layer` is updated (e.g. from 'below' to 'above' back to 'below').

etpinard commented on src/components/shapes/index.js:
```diff
@@ -288,23 +294,72 @@ function updateShape(gd, index, opt, value) {
+        });
+    }
+
+    return;
```
:cow2: no need for this. Save the bytes.

n-riesco replied:
I will remove it.

I usually write closures at the end of a function and use the `return` to indicate that there are no more statements hidden after the closures.

etpinard replied:
I understand. I'm not saying that it's a bad habit. It's really a matter of consistency with the rest of the source code.

etpinard commented on src/components/shapes/index.js:
```diff
@@ -288,23 +294,72 @@ function updateShape(gd, index, opt, value) {
+    return shape.layer === 'below' && (xa === shape.xref || ya === shape.yref);
+}
+
+function forEachSubplot(gd, fn) {
```
You may argue that adding `forEachSubplot` makes things *more readable*, but scrolling to the bottom of the page to see what a function does is anything but *reader-friendly*. 

Please move the `for` loop below in the `else` block on [line 310](https://github.com/plotly/plotly.js/pull/439/files#diff-a123bebe752778fa4640d499ad73277cR310).

etpinard commented on src/components/shapes/index.js:
```diff
@@ -288,23 +294,72 @@ function updateShape(gd, index, opt, value) {
+
+function forEachSubplot(gd, fn) {
+    var plots = gd._fullLayout._plots || {},
+        subplots = Object.getOwnPropertyNames(plots);
```
:cow2: Agreed. `getOwnPropertyNames` is safer, but `fullLayout._plots` should not have anything but *own properties*.

So, please swap `Object.getOwnPropertyNames` for `Object.keys` for consistency with the rest of the code (and to save a few bytes :smiley: )

etpinard commented on test/jasmine/tests/shapes_test.js:
```diff
@@ -24,30 +24,114 @@ describe('Test shapes:', function() {
+                .toEqual(countSubplots(gd));
+        });
+
+        it('has as many *path* nodes as shapes in the subplot', function() {
```
great tests! :beers:

etpinard commented on test/jasmine/tests/shapes_test.js:
```diff
@@ -75,33 +159,32 @@ describe('Test shapes:', function() {
-                expect(countShapeLayers()).toEqual(1);
-                expect(countShapePaths()).toEqual(pathCount + 1);
+            Plotly.relayout(gd, 'shapes[' + index + ']', shape).then(function()
+            {
```
we should really pick an `eslint` [brace-style](http://eslint.org/docs/rules/brace-style).

I'd vote for `Stroustrup`, I know @mdtusz would vote for `1tbs`, but no one on our would vote for `Allman`.

So please, bring up that `{` on that same line as `function()`.

n-riesco replied:
I'm trying to keep lines 80 chars or less. Shall I break the line before the function, like this:

```
            Plotly.relayout(gd, 'shapes[' + index + ']', shape).then(
                function() {
                    expect(countShapePathsInUpperLayer())
                        .toEqual(pathCount + 1);
                    expect(getLastShape(gd)).toEqual(shape);
                    expect(countShapes(gd)).toEqual(index + 1);
                }).then(done);
```

etpinard replied:
Ah good point. Don't worry about going over 80 characters. I don't think we'll ever enforce that rule in our jasmine test suites.

etpinard commented on test/jasmine/tests/shapes_test.js:
```diff
@@ -75,33 +159,32 @@ describe('Test shapes:', function() {
-                expect(countShapeLayers()).toEqual(1);
-                expect(countShapePaths()).toEqual(pathCount + 1);
+            Plotly.relayout(gd, 'shapes[' + index + ']', shape).then(function()
+            {
```
:cow2: again, bring up `{`

etpinard commented on test/jasmine/tests/shapes_test.js:
```diff
@@ -24,30 +24,114 @@ describe('Test shapes:', function() {
+    }
+
+    function countSubplots(gd) {
+        return Object.getOwnPropertyNames(gd._fullLayout._plots || {}).length;
```
:cow2: `Object.keys`

etpinard commented on src/components/shapes/index.js:
```diff
@@ -288,23 +294,72 @@ function updateShape(gd, index, opt, value) {
+    }
+}
+
+function getShapeLayer(gd, index) {
```
nice abstraction here. :beers:

harika-strand replied:
Is this thing possible in 3d ???

n-riesco commented:
@etpinard I've addressed all the points in [the TODO list](https://github.com/plotly/plotly.js/pull/439#issuecomment-211553043) and fixed a bug that prevented shape properties from being updated.

I've added [your plot](https://github.com/plotly/plotly.js/pull/439#issuecomment-211552620) to the image mocks, because it nicely exercises all the new functionality, i.e.:
- places shapes in the lower, upper and subplot shape  layers,
- and uses shapes that extend across several subplots (to test the appropriate clip paths are used).

etpinard commented on test/jasmine/tests/shapes_test.js:
```diff
@@ -186,5 +186,45 @@ describe('Test shapes:', function() {
+                    .toEqual(shapesInUpperLayer + 1);
+                expect(getLastShape(gd)).toEqual(shape);
+                expect(countShapes(gd)).toEqual(index + 1);
+            }).then(done);
```
nicely done

etpinard commented on src/components/shapes/index.js:
```diff
@@ -298,25 +298,33 @@ function updateShape(gd, index, opt, value) {
+            subplots = Object.keys(plots),
+            plotinfo;
+
+        for(i = 0, n = subplots.length; i < n; i++) {
```
thank you very much

etpinard commented:
@n-riesco great PR! :beers:

rolandjitsu commented:
I've got a question: When using shapes, if I set layer `below`, the grid lines are above the layer, is this expected behaviour? If yes, is there a way to place the shapes under the traces but above the grid lines?
